pipelines/oracle_grounded/parity_view_sets.py

```
@@ -0,0 +1,157 @@
+
+def _dropped_record_errors(record_ids, view_id_counts, where):
+    """Every source record must still have a view."""
+    # Only string record IDs are hashable-safe to check against view_id_counts;
+    # a malformed ID (e.g. a list) is already reported by the validity pass and
+    # must not reach `in` on a dict, which raises TypeError for unhashable keys.
+    dropped = [rid for rid in record_ids if isinstance(rid, str) and rid not in view_id_counts]
+    if not dropped:
+        return []
+    return [
+        f"{where}: training view set drops records {dropped} "
+        f"[TRAINING_VIEW_HIDES_FAILURE]"
```

Ruff (0.16.3)

[warning] Unparenthesized implicit string concatenation in collection

Did you forget a comma?

(ISC004)